atlasdb-client/src/test/java/com/palantir/atlasdb/stream/BlockConsumingInputStreamTest.java

```diff
@@ -0,0 +1,239 @@
+        byte[] result = new byte[DATA_SIZE_PLUS_ONE];
+        int read = stream.read(result);
+        assertEquals(DATA_SIZE, read);
+        assertArrayEquals(data, Arrays.copyOf(result, DATA_SIZE));
+    }
+
+    @Test
+    public void can_read_multiple_blocks() throws IOException {
+        BlockConsumingInputStream stream = BlockConsumingInputStream.create(singleByteConsumer, DATA_SIZE, 1);
+        byte[] result = new byte[DATA_SIZE];
+        int read = stream.read(result);
+        assertEquals(DATA_SIZE, read);
+        assertArrayEquals(data, result);
+    }
+
+    @Test
+    public void can_read_multiple_blocks_and_partially_fill_result() throws IOException {
+        BlockConsumingInputStream stream = BlockConsumingInputStream.create(singleByteConsumer, DATA_SIZE, 1);
+        byte[] result = new byte[DATA_SIZE_PLUS_ONE];
+        int read = stream.read(result);
+        assertEquals(DATA_SIZE, read);
+        assertArrayEquals(data, Arrays.copyOf(result, DATA_SIZE));
+    }
+
+    @Test(expected = IndexOutOfBoundsException.class)
+    public void passing_in_too_many_blocks_causes_an_exception() throws IOException {
+        BlockConsumingInputStream stream = BlockConsumingInputStream.create(singleByteConsumer, DATA_SIZE_PLUS_ONE, 1);
+        byte[] result = new byte[DATA_SIZE_PLUS_ONE];
+        //noinspection ResultOfMethodCallIgnored
+        stream.read(result);
+    }
+
+    @Test
+    public void passing_in_too_few_blocks_causes_incomplete_output() throws IOException {
+        int dataSizeMinusOne = 3;
+        BlockConsumingInputStream stream = BlockConsumingInputStream.create(singleByteConsumer, dataSizeMinusOne, 1);
+        byte[] result = new byte[DATA_SIZE];
+        int read = stream.read(result);
+        assertEquals(dataSizeMinusOne, read);
+        assertArrayEquals(Arrays.copyOf(data, dataSizeMinusOne), Arrays.copyOf(result, dataSizeMinusOne));
+    }
+
+    @Test
+    public void can_load_multiple_blocks_at_once_and_also_fewer_blocks_at_end() throws IOException {
+        BlockGetter spiedGetter = Mockito.spy(new MockableBlockGetter(singleByteConsumer));
+        BlockConsumingInputStream stream = BlockConsumingInputStream.create(spiedGetter, DATA_SIZE, 3);
+        //noinspection ResultOfMethodCallIgnored
+        stream.read();
+        verify(spiedGetter, times(1)).get(anyInt(), eq(3), any());
+
+        byte[] ata = new byte[3];
+        int bytesRead = stream.read(ata);
+        assertEquals(3, bytesRead);
+        verify(spiedGetter, times(1)).get(anyInt(), eq(1), any());
+    }
+
+    private void expectNextBytesFromStream(BlockConsumingInputStream stream, String expectedOutput) throws IOException {
+        byte[] chunk = new byte[2];
+        int read = stream.read(chunk);
+        assertEquals(2, read);
+        assertArrayEquals(expectedOutput.getBytes(), chunk);
+    }
+
+    private class MockableBlockGetter implements BlockGetter {
```

Comment:
Why was this necessary? You should be able to spy on the `singleByteGetter` directly.